src/dftbp/timedep/linrespcommon.F90

```diff
@@ -689,15 +703,381 @@ subroutine actionAplusB(spin, wij, sym, win, nocc_ud, nvir_ud, nxoo_ud, nxvv_ud,
+    type(TTransCharges), intent(in) :: transChrg
+
+    !> vector to multiply with size(nmat)
+    real(dp), intent(in) :: vin(:)
+
+    !> vector containing the result on exit size(nmat)
+    real(dp), intent(out) :: vout(:)
+
+    !> is calculation range-separated?
+    logical, intent(in) :: tRangeSep
+
+    !> long-range Gamma if in use
+    real(dp), allocatable, optional, intent(in) :: lrGamma(:,:)
+
+    integer :: nmat, natom, norb, ia, nxvv_a, nSpin
+    integer :: ii, aa, ss, jj, bb, ias, jbs, abs, ijs, jas, ibs
+    integer :: iam, nLoc, myia, myab, myja, ierr, iGlb, fGlb, nProcs, nLocAB
+    integer, allocatable ::  getABasym(:,:), locSizeAB(:), vOffsetAB(:)
+
+    ! somewhat ugly, but fast small arrays on stack:
+    real(dp) :: otmp(size(gamma, dim=1)), gtmp(size(gamma, dim=1))
+    real(dp) :: qij(size(gamma, dim=1)) ! qij Working array (used for excitation charges). (nAtom)
+    real(dp) :: vout_ons(size(vin))
+    real(dp), allocatable :: vLoc(:), vGlb(:), vGlb2(:)
+    real(dp), dimension(2) :: spinFactor = (/ 1.0_dp, -1.0_dp /)
+    ! for later use to change HFX contribution
+    real(dp), allocatable :: qv(:,:)
+
+    @:ASSERT(size(vin) == size(vout))
+
+    !! dimension guessed from input vector
+    nmat = size(vin)
+    @:ASSERT(nmat <= nxov_rd)
+    natom = size(gamma, dim=1)
+    norb = nocc_ud(1)+nvir_ud(1)
+    
+    iam = env%mpi%globalComm%rank
+    nLoc = locSize(iam+1)
+    @:ASSERT(nmat == nLoc)
+    iGlb = vOffset(iam+1) + 1 
+    fGlb = vOffset(iam+1) + nLoc
+    allocate(vLoc, mold=vin)
+    vLoc = 0.0_dp
+    vout = 0.0_dp
+
+    if(tAplusB) then
+      vLoc(:) = vin
+    else
+      vLoc(:) = vin * sqrt(wij(iGlb:fGlb))
+    endif
+
+    !> Compute w_ia = q^ia_A G_AB q^jb v_jb 
+
+    oTmp(:) = 0.0_dp    
+    do myia = 1, nLoc
+      ia = vOffset(iam+1) + myia
+      qij(:) = transChrg%qTransIA(ia, env, denseDesc, ovrXev, grndEigVecs, getIA, win)
+      otmp(:) = otmp(:) + qij(:) * vLoc(myia) * sqrOccIA(ia)
+    enddo
+
+    call mpifx_allreduceip(env%mpi%globalComm, otmp, MPI_SUM)
+
+    if (.not.spin) then !-----------spin-unpolarized systems--------------
+
+      if (sym == 'S') then
+
+        call hemv(gtmp, gamma, otmp)
+
+        ! 4 * wn * (g * Q)
+        vOut(:) = 0.0_dp
+        do myia = 1, nLoc
+          ia = vOffset(iam+1) + myia
+          qij(:) = transChrg%qTransIA(ia, env, denseDesc, ovrXev, grndEigVecs, getIA, win)
+          vOut(myia) = 4.0_dp * sqrOccIA(ia) * dot_product(qij, gTmp)
+        enddo  
+
+      else
+       
+        otmp = otmp * spinW(species0)
+
+        ! 4 * wn * (o * Q)
+        vOut = 0.0_dp
+        call transChrg%qVecMat(env, denseDesc, ovrXev, grndEigVecs, getIA, win, oTmp, vOut, vOffset(iam+1))
+        vOut(:) = 4.0_dp * sqrOccIA(iGlb:fGlb) * vOut
+
+
+      end if
+
+    else !--------------spin-polarized systems--------
+
+      call hemv(gtmp, gamma, otmp)
+
+      otmp(:) = 0.0_dp
+
+      !$OMP PARALLEL DO DEFAULT(SHARED) PRIVATE(ia,ss,qij) &
```

Comment:
Does OMP make sense inside a MPI region of code?

Reply:
You don't think it will be executed OMP?

Reply:
If it's executed as MPI with current structures, probably not at the moment. But if in future we start using node level parallelism, possibly.